File: Leetcode 3401-3500/3404/3404.java
// Leetcode 3404: Count Special Subsequences
// https://leetcode.com/problems/count-special-subsequences/
// Solved on 4th of December, 2025
import java.util.HashMap;
import java.util.Map;

class Solution {
    /**
     * Counts the number of special subsequences in the given array.
     *
     * @param nums The input array of integers.
     * @return The total count of special subsequences.
     */
    public long numberOfSubsequences(int[] nums) {
        long count = 0;
        int n = nums.length;
        Map<Double, Integer> freq = new HashMap<>();
        for (int r = 4; r < n - 2; r++) {
            int q = r - 2;
            for (int p = 0; p < q - 1; p++) {
                double ratio = (double) nums[p] / nums[q];
                freq.put(ratio, freq.getOrDefault(ratio, 0) + 1);
            }
            for (int s = r + 2; s < n; s++) {
                double ratio = (double) nums[s] / nums[r];
                count += freq.getOrDefault(ratio, 0);
            }
        }
        return count;
    }
}

File: Leetcode 3401-3500/3404/3404.py
# Leetcode 3404: Count Special Subsequences
# https://leetcode.com/problems/count-special-subsequences/
# Solved on 4th of December, 2025
from collections import defaultdict
from math import gcd


class Solution:
    def numberOfSubsequences(self, nums: list[int]) -> int:
        """
        Counts the number of special subsequences in the given list of integers.

        Args:
            nums: A list of integers.
        Returns:
            The total count of special subsequences.
        """
        n = len(nums)
        ans = 0
        ratioCnt = defaultdict(int)

        for r in range(4, n - 2):
            q = r - 2
            for p in range(q - 1):
                valP = nums[p]
                valQ = nums[q]
                common = gcd(valP, valQ)
                key = (valP // common, valQ // common)
                ratioCnt[key] += 1

            valR = nums[r]
            for s in range(r + 2, n):
                valS = nums[s]
                common = gcd(valS, valR)
                key = (valS // common, valR // common)
                ans += ratioCnt[key]

        return ans